### ISSUE TITLE
k8s-openapi memory consumption blocks CI

### ISSUE THREAD

**Issue by clux** [clux/kube-rs#62]
Did a full `cargo test --all-features` after having bumped the `k8s-openapi` dependency and memory use is almost 8GB with parallelism. Looks like it takes >4GB just compiling the dependency itself. This blocks CircleCI from building it (on the free account), and travis just seems to time out. Need a better CI solution..

**Comment by clux** [comment]
God, azure pipelines is clunky. ...BUT, it does have the required amount of memory.

**Comment by clux** [comment]
Literally unable to even compile k8s-openapi at all now on circleci. Wonder if it's possible to generate a subset of the openapi types without all the api stuff  around it :thinking:

**Comment by Arnavion** [comment]
Try https://github.com/Arnavion/k8s-openapi/blob/master/ci/k8s-openapi.sh#L19-L20

Otherwise let me know if you have any suggestions. Both the repo and $dayjob use Azure DevOps, so I haven't noticed any issues.

A default-enabled feature for API could work, though I dislike default-enabled features (they usually become enabled by other deps in the dep graph without any way to disable them).

**Comment by clux** [comment]
Thanks for the suggestion. Unfortunately it doesn't appear to have a significant impact on a [free circle job](https://circleci.com/gh/clux/kube-rs/15).

I agree that while non-default features certainly require some care to use, you can usually work around that for libraries that aren't used everywhere in the dep tree.

Do you think k8s-openapi will be pulled in by a lot more than a core controller and possibly this library? If it's only to be used in a few places, an optionally lighter version would still be very helpful in terms of making this buildable on free CI.

As for actual suggestions for implementing it, I'd have to dig some more into your generation process first. I am still confused to why it consumes this much tbh, it was fine in 0.4.0, and it's clearly very clever in how it's only including the version you want without that extra code affecting your chosen feature build.

**Comment by Arnavion** [comment]
>Thanks for the suggestion. Unfortunately it doesn't appear to have a significant impact on a [free circle job](https://circleci.com/gh/clux/kube-rs/15).

Drat. I don't have any other ideas.

>I agree that while non-default features certainly require some care to use, you can usually work around that for libraries that aren't used everywhere in the dep tree.
>
>Do you think k8s-openapi will be pulled in by a lot more than a core controller and possibly this library? If it's only to be used in a few places, an optionally lighter version would still be very helpful in terms of making this buildable on free CI.

I don't know.

I'll think about adding a feature for the API functions. Mostly my dislike of default-enabled features comes from my experience from maintaining `derive-error-chain`, but it's true that `k8s-openapi` won't necessarily have the same problem.

>I am still confused to why it consumes this much tbh, it was fine in 0.4.0, and it's clearly very clever in how it's only including the version you want without that extra code affecting your chosen feature build.

Yes. If anything 0.5.0 has fewer parameter structs due to the merging of `{Delete,List,Patch,Watch}Optional`, and many of the response enums are smaller because they dropped their zero-sized variants. I'm not sure why it would take more resources.

I see that you only updated to 0.5.0 two weeks ago, so it also shouldn't be something like a rustc regression that coincides with using 0.5.0.

I do see that you also switched from `v1_13` to `v1_14`. It might be useful to see if that is the reason, in case `v1_14` has a lot more types compared to `v1_13`.

**Comment by clux** [comment]
Oh, good hunch! Downgrading it to `v1_13 ` did actually manage to push it under 4G! It might just be the net-worth of structs then. Looking at it during compile it does look like the memory consumption peaks at the end of of compilation, maybe it's possible to tell cargo to be more clever about this. Will try more tomorrow.

**Comment by Arnavion** [comment]
Strange. There's not that much difference between `v1_13` and `v1_15` in terms of new structs:

```sh
$ diff <(cd v1_13; find -name '*.rs') <(cd v1_15; find -name '*.rs')

2,6d1
< ./api/admissionregistration/v1alpha1/initializer.rs
< ./api/admissionregistration/v1alpha1/initializer_configuration.rs
< ./api/admissionregistration/v1alpha1/initializer_configuration_list.rs
< ./api/admissionregistration/v1alpha1/mod.rs
< ./api/admissionregistration/v1alpha1/rule.rs
7a3
> ./api/admissionregistration/v1beta1/mutating_webhook.rs
11a8
> ./api/admissionregistration/v1beta1/validating_webhook.rs
14d10
< ./api/admissionregistration/v1beta1/webhook.rs
223a220,223
> ./api/coordination/v1/lease.rs
> ./api/coordination/v1/lease_list.rs
> ./api/coordination/v1/lease_spec.rs
> ./api/coordination/v1/mod.rs
260a261
> ./api/core/v1/csi_volume_source.rs
407a409
> ./api/core/v1/windows_security_context_options.rs
413a416
> ./api/extensions/v1beta1/allowed_csi_driver.rs
462a466
> ./api/extensions/v1beta1/runtime_class_strategy_options.rs
478a483,500
> ./api/networking/v1beta1/http_ingress_path.rs
> ./api/networking/v1beta1/http_ingress_rule_value.rs
> ./api/networking/v1beta1/ingress.rs
> ./api/networking/v1beta1/ingress_backend.rs
> ./api/networking/v1beta1/ingress_list.rs
> ./api/networking/v1beta1/ingress_rule.rs
> ./api/networking/v1beta1/ingress_spec.rs
> ./api/networking/v1beta1/ingress_status.rs
> ./api/networking/v1beta1/ingress_tls.rs
> ./api/networking/v1beta1/mod.rs
> ./api/node/mod.rs
> ./api/node/v1alpha1/mod.rs
> ./api/node/v1alpha1/runtime_class.rs
> ./api/node/v1alpha1/runtime_class_list.rs
> ./api/node/v1alpha1/runtime_class_spec.rs
> ./api/node/v1beta1/mod.rs
> ./api/node/v1beta1/runtime_class.rs
> ./api/node/v1beta1/runtime_class_list.rs
479a502
> ./api/policy/v1beta1/allowed_csi_driver.rs
495a519
> ./api/policy/v1beta1/runtime_class_strategy_options.rs
538a563,565
> ./api/scheduling/v1/mod.rs
> ./api/scheduling/v1/priority_class.rs
> ./api/scheduling/v1/priority_class_list.rs
566a594,600
> ./api/storage/v1beta1/csi_driver.rs
> ./api/storage/v1beta1/csi_driver_list.rs
> ./api/storage/v1beta1/csi_driver_spec.rs
> ./api/storage/v1beta1/csi_node.rs
> ./api/storage/v1beta1/csi_node_driver.rs
> ./api/storage/v1beta1/csi_node_list.rs
> ./api/storage/v1beta1/csi_node_spec.rs
612a647
> ./apimachinery/pkg/apis/meta/v1/fields.rs
618a654
> ./apimachinery/pkg/apis/meta/v1/managed_fields_entry.rs
```

Sure there are a bunch of new types with their own operation functions, but I wouldn't think they'd be enough to increase the compiler's memory usage from 4 to 8 GiB...

**Comment by Arnavion** [comment]
I ran a couple of tests of taking https://github.com/clux/kube-rs/commit/04ea9256e78cd51bb58974acc20731fa94556eff and https://github.com/clux/kube-rs/commit/2d9b00c50323428bd586edc75c133ed7e688dc6e and running `cargo test --lib` like your CI did.

I don't see much difference in the RSS used - ~4.4 GiB for `v1_13` vs ~4.5GiB for `v1_15` (as measured with `/usr/bin/time -v` and from eyeballing the `rustc` process in `htop`). `v1_13` always uses slightly less RSS than `v1_15`, but it's nothing as dramatic as 4 GiB vs 8 GiB.

I also don't see any difference between Rust stable and nightly in this regard. And I get identical numbers with the `k8s-openapi-tests` crate in my repo.

I suspect it was always close to hitting the CircleCI agent's limit, and the ~100 MiB from the 30 extra structs and API operations was enough to push it over the limit.

**Comment by Arnavion** [comment]
Try `k8s-openapi = { git = "https://github.com/Arnavion/k8s-openapi", branch = "feature-api", default-features = false }`. I see the RSS max out at ~1.4 GiB when building `k8s-openapi-tests`. If it works for you I'll see about releasing it as v0.5.1

**Comment by clux** [comment]
Woahh, this is fantastic! Yes, this is a huge improvement. Memory usage is now a blip, and build times also seem a lot faster. Ran through all the examples and everything seems to work fine on this end.

**Comment by clux** [comment]
(I was hitting one limit still when linking in 8 different executables using k8s-openapi at the same time when doing a full test with --all-features, but that's more a sanity thing that I could tweak by limiting that particular instruction to `-j4`. https://github.com/clux/kube-rs/commit/925281b8573162fbe64fd76d173f0ab9f282d57c ).

Am able to build everything on circle easily now. Thank you so much!

**Comment by clux** [comment]
> I suspect it was always close to hitting the CircleCI agent's limit, and the ~100 MiB from the 30 extra structs and API operations was enough to push it over the limit.

Yeah, that seems entirely possible. I was seeing travis-ci just freeze during the last k8s-openapi bump and we could have hit that from just a single build then.

The implicit parallelism on linking locally + on circle made it seem even worse (that's how I saw the 8GB number locally).

**Comment by Arnavion** [comment]
I tried a few other approaches like [moving most of the API functions' bodies into a single common function](https://github.com/Arnavion/k8s-openapi/compare/build-request-common-fn) and removing the `Debug` impl on `Optional` types but it made no difference. It really does look like the number of functions / types is the cause of rustc's memory usage.

I've added the `api` feature gating to [master](https://github.com/Arnavion/k8s-openapi/commit/af3e4d1430a3e5346f8774de80f35f0674d23dfc) now. I'll release it as v0.5.1 some time this week.

Long term I'm thinking of combining more of the `Optional` types into single global ones like I did for `{Delete,List,Patch,Watch}Optional`. Might be worth combining some of the `Response` types too. (These are much less clear-cut - eg `read_namespaced_pod`, `read_namespaced_pod_log` and `read_namespaced_pod_status` are all `"x-kubernetes-action": "get"` operations associated with `/v1/Pod` but have disparate parameters.)

**Comment by clux** [comment]
Yeah, if re-use of types are possible that ought to help, but it's non-trivial, and bound to miss some cases if you over do it (like this library). I guess this is just one enormous API.

Anyway, I really appreciate all the work you've done here for this. It looks daunting to work with that many feature-based `if` statements, but it's certainly made the dependency a lot lighter. Thank you.

**Comment by Arnavion** [comment]
I've released v0.5.1

**Comment by clux** [comment]
Thank you very much. Updated all references to it in here :slightly_smiling_face:
